File: Project1/src/project1/animals/Animal.java
package project1.animals;

public abstract class Animal {
  protected String name;
  /**
   *  Abstract methods
   */
  public abstract void eat();
  public abstract void makeNoise();

  /**
   *  Prints that the animal in now roaming
   */
  public void wakeUp(){
    System.out.println(this.toString() + " is now awake");
  }

  /**
   *  Prints that the animal in now roaming
   */
  public void roam(){
    System.out.println(this.toString() + " is now roaming");
  }

  /**
   *  Prints that the animal in now asleep
   */
  public void sleep(){
    System.out.println(this.toString() + " went to sleep");
  }

  /**
   *  @param name String
   *  Name of the animal
   */
  public Animal(String name){
    this.name = name;
  }

  /**
   *  @return String
   *  Name of the animal
   */
  public String getName(){
    return this.name;
  }

  /**
   *  @return String
   *  Returns string of the type of species
   */
  public String getSpecies(){
    return this.getClass().getSimpleName();
  }

  /**
   *  @return String
   *  Name and species of animal
   */
  public String toString(){
    return this.name + " the " + this.getClass().getSimpleName();
  }
}

File: Project2/src/animals/Animal.ts
import { Noise } from "./";

export default abstract class Animal {
  private name: String;
  private noise: Noise;

  /**
   *  @param name String
   *  Name of the animal
   *  @param noise Noise
   *  The Noise object that the animal will make
   */
  public constructor(name: String, noise: Noise){
    this.name = name;
    this.noise = noise;
  }

  /**
   *  Abstract methods
   */
  public abstract eat();

  public makeNoise(){
    if(this.noise != null)
      console.log(this.toString() + " " + this.noise.getNoise());
    else // If noise is null, do default behavior
      console.log(this.toString() + " is silent");
  }

  /**
   *  Prints that the animal in now roaming
   */
  public wakeUp(){
    console.log(this.toString() + " is now awake");
  }

  /**
   *  Prints that the animal in now roaming
   */
  public roam(){
    console.log(this.toString() + " is now roaming");
  }

  /**
   *  Prints that the animal in now asleep
   */
  public sleep(){
    console.log(this.toString() + " went to sleep");
  }

  /**
   *  @return String
   *  Name of the animal
   */
  public getName(){
    return this.name;
  }

  /**
   *  @return String
   *  Returns string of the type of species
   */
  public getSpecies(){
    var funcNameRegex = /function (.{1,})\(/;
    var results = (funcNameRegex).exec((<any> this).constructor.toString());
    return (results && results.length > 1) ? results[1] : "";
  }

  /**
   *  @return String
   *  Name and species of animal
   */
  public toString(){
    return this.name + " the " + this.getSpecies();
  }
}
